src/placement/pl_map.c

```diff
@@ -631,8 +633,19 @@ pl_select_leader(daos_obj_id_t oid, uint32_t grp_idx, uint32_t grp_size,
 		while (shard->po_rebuilding || shard->po_shard == -1 ||
 		       shard->po_target == -1) {
 			idx--;
-			if (++fail_cnt >= oc_attr->u.ec.e_p)
-				return -DER_IO;
+			if (shard->po_rebuilding)
+				parity_rebuilding = true;
+
+			if (++fail_cnt >= oc_attr->u.ec.e_p) {
+				/* If parity is rebuilding, let's return DER_STALE,
+				 * so object I/O might refresh the pool map and layout
+				 * until parity rebuilt finish.
+				 */
+				if (parity_rebuilding)
+					return -DER_STALE;
+				else
+					return -DER_IO;
```

Comment:
this seems can cause a problem for example, 4+2 EC obj, firstly only lost one parity shard, then in-rebuilding that parity shard. Suddenly it lost 2 other data shards, that will cause the rebuild fail and cannot finish, and then here also return DER_STALE and client IO retry forever.
To avoid that, seems need some other check?

Reply:
Hmm, if there is one parity left, it will always choose the available parity as the leader, i.e. return 0 here.

If both parity fail, then all parity are in rebuilding, then it will return STALE and retry here indeed.  So if there are further data shards failed, then rebuild will fail anyway, and po_rebuilding will be reset to 0 and po_target/po_shard will be set to -1.

or I misunderstand here.

Reply:
Ok, seems fine. Just one possible refine later, that if #failures > p, then can return EIO, rather than wait rebuild finish and update pool map